### ISSUE TITLE
Prompt user for just one broken business rule

### ISSUE THREAD

**Issue by Chicagoan2016** [MarimerLLC/cslaforum#800]
Hi all,
This may be a simple question but I want to prompt the user only for a certain broken business rule, e.g. consider Name property of Person class, the name property has 'required' business rule, 'ToUpper' business rule and 'CertainLength' business rule, if I want to prompt the user for cerainlength business rule, how would I go about it?

Kind regards

**Version and Platform**
CSLA version: 4.7.100
OS: Windows, Linux, iOS, Android, etc.
Platform: WinForms, WPF, ASP.NET Core, MVC, Xamarin, etc.

**Comment by michaelcsikos** [comment]
You normally only display one at a time, based on the priority. If the `Required` rule has a higher priority that `CertainLength`, `Required` will show until there's a value, then the next broken rule will appear. You could make one custom rule class to check everything, and display them all together.

**Comment by rockfordlhotka** [comment]
What do you mean by "prompt"?

As in "display" - show the broken rule to the user?

Or as in "pop up a modal dialog" about the broken rule?

**Comment by Chicagoan2016** [comment]
Thank you @rockfordlhotka , @michaelcsikos , in the example above, I have given the 'Required' rule a priority of 0, ToUpper a priority of 1, and 'CertainLegth' has a priority of 2. I only want to show the user a specific dialog when the 'CertainLength' business rule is broken.
I have to use a trivial example as I can't use example from production code : ).

Kind Regards

**Comment by hurcane** [comment]
This sounds to me like a UI concern. If you're using MVVM, I would add behavior in the ViewModel to examine the broken rules collection, and then take appropriate action if the specific rule is broken.

**Comment by Chicagoan2016** [comment]
@michaelcsikos , you are right I have to take care of this in the UI, I don't know how to access a specific business rule from broken business rules collection : )

**Comment by rockfordlhotka** [comment]
Use a LINQ query against the collection?

**Comment by fujiiface** [comment]
That’s how I would do it.  Query the BrokenRulesCollection of the object.

**Comment by Chicagoan2016** [comment]
@rockfordlhotka , @fujiiface , thank you, I guess I was too close to the project : ), I came up with something like below
`
 var mybrokenrule = from r in objPerson.BrokenRulesCollection
                           where r.Property == Person.CertainLength.Name
                                && new RuleUri(r.RuleName).RuleTypeName.ToLower().Split('.').Last() == nameof(CertainLengthRequired).ToLower()
                            select r;
`

does it look right? kinda looks 'convoluted' 


kinds regards

**Comment by rockfordlhotka** [comment]
That looks right, yes.

**Comment by Chicagoan2016** [comment]
Thank you @rockfordlhotka , as the csla objects implement IDataErrorInfo interface,  where should the above code go? in validated or validating event?
Kind regards

**Comment by rockfordlhotka** [comment]
I haven't done any Windows Forms for so long, I don't remember where to put that code.

You need to put it somewhere that you can interact with the user and get the desired UI workflow.

**Comment by Chicagoan2016** [comment]
Thank you @rockfordlhotka

**Comment by Chicagoan2016** [comment]
uh oh, we just found out that after finding the 'certain'  broken business rule, we prompt the user to correct the issue and after the user is done we need to run the 'certain' business rule again to make sure the object is valid ( so we could save the object).
Any ideas please ?

**Comment by rockfordlhotka** [comment]
Does the user not fix the issue by editing a property? That should trigger running any property-specific  rules automatically.

**Comment by Chicagoan2016** [comment]
I couldn't give the example of production code and my example code doesn't portray the use case accurately :).
Let me try another example, we have an Employee object and one property of the object is 'JobTitle'
JobTitle property is set from a combobox where the combobox is populated by a read only collection, say, JobTileList.cs.
When a user enter information for an employee and picks a JobTitle that is already in the combobox (and in the database) everything is fine, but we allow the users to enter a new JobTitle, a  JobTitle that doesn't exist in the combobox (and in DB), this 'newJobTitle' rule is implemented in Person object and this is where we prompt the user if she wants to save the new JobTitle, if she picks 'Yes' we add the new JobTitle to the database (using a Command object and invalidating the cache for JobTitleList).
Since the value entered by the user propagates to the Person object (using databinding in winforms), the business rule associated with new "JobTitle" doesn't run **again** because even when the user picks 'Yes' the value of the JobTitle remains the same in Person object.

I hope i made sense : )

**Comment by hurcane** [comment]
The business of adding and validating the JobTitle record does not belong to Person. Your UI is maintaining two different business objects, a list of Job titles and a Person.

The combo should be bound to and interact with the editable business list of Job titles.

The UI (ViewModel, Controller, form backing code, pick your flavor) assigns the Job title to the Person class, if the Job title has no broken rules. If the job title is not invalid, you might want to clear the property from the Person class.
